common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java

```diff
@@ -128,37 +136,60 @@ private void tryCastSpell(SpellUnit a, SpellUnit b, SpellUnit c) {
         List<SpellDirection> spell = Stream.of(a, b, c)
                 .map(x -> (x == SpellUnit.PRIMARY) != isSpellInverted ? SpellDirection.RIGHT : SpellDirection.LEFT)
                 .toList();
-        for (int i = 0; i < spellInProgress.length; ++i) {
-            if (spellInProgress[i] != spell.get(i)) {
-                sendCancelReason(Component.translatable("feature.wynntils.quickCast.incompatibleInProgress"));
-                return;
+
+        if (safeCasting.get()) {
+            for (int i = 0; i < spellInProgress.length; ++i) {
+                if (spellInProgress[i] != spell.get(i)) {
+                    sendCancelReason(Component.translatable("feature.wynntils.quickCast.incompatibleInProgress"));
+                    return;
+                }
             }
         }
 
         lastSelectedSlot = McUtils.inventory().selected;
         List<SpellDirection> remainder = spell.subList(spellInProgress.length, spell.size());
-        remainder.stream().map(SpellDirection::getSendPacketRunnable).forEach(SPELL_PACKET_QUEUE::add);
+        SPELL_PACKET_QUEUE.addAll(remainder);
     }
 
     @SubscribeEvent
     public void onTick(TickEvent e) {
         if (!Models.WorldState.onWorld()) return;
 
         // Clear spell after the 40 tick timeout period
-        if (spellCountdown > 0 && --spellCountdown <= 0) spellInProgress = SpellDirection.NO_SPELL;
+        if (spellCountdown > 0 && --spellCountdown <= 0) {
+            spellInProgress = SpellDirection.NO_SPELL;
+        }
 
         if (SPELL_PACKET_QUEUE.isEmpty()) return;
         if (--packetCountdown > 0) return;
 
         int currSelectedSlot = McUtils.inventory().selected;
         boolean slotChanged = currSelectedSlot != lastSelectedSlot;
 
-        if (slotChanged) McUtils.sendPacket(new ServerboundSetCarriedItemPacket(lastSelectedSlot));
-        SPELL_PACKET_QUEUE.poll().run();
-        if (slotChanged) McUtils.sendPacket(new ServerboundSetCarriedItemPacket(currSelectedSlot));
+        if (slotChanged) {
+            McUtils.sendPacket(new ServerboundSetCarriedItemPacket(lastSelectedSlot));
+        }
+
+        // Right clicks need a tick delay between them (2-3 ticks), left clicks don't
+        boolean didRightClick = false;
+        do {
+            SpellDirection spellDirection = SPELL_PACKET_QUEUE.poll();
+            spellDirection.getSendPacketRunnable().run();
+
+            if (spellDirection == SpellDirection.RIGHT) {
+                didRightClick = true;
+            }
+        } while (!SPELL_PACKET_QUEUE.isEmpty()
+                && (!didRightClick || SPELL_PACKET_QUEUE.peek() != SpellDirection.RIGHT));
+
+        if (slotChanged) {
+            McUtils.sendPacket(new ServerboundSetCarriedItemPacket(currSelectedSlot));
+        }
 
         // Waiting a few ticks is useful for avoiding lag related input-overlaps
```

Comment:
This packet tick pause is only used between two consecutive right clicks, or after the last click of the spell. Am I reading this right? If so, the comment could perhaps be updated to reflect that.

Reply:
Yes. Okay. I think this can be even faster, we only need this for right-clicks.

Reply:
So if you have like a "LRL, LRL" sequence, you would not need a single pause at all?

Reply:
Well, you can't do the two Rs in one tick. I'll just update the comment and keep this logic..

Reply:
Can we replace all this with some kind of "click sender" machinery, that has a simple state machine that knows if the last click sent was a right click, and if so and the new click to send is a right click, it adds a pause?

Reply:
Like `ClickSendHandler`?

Reply:
We probably can, but this PR was just meant to be a quick fix. That new system would need work + moderation checks.

Reply:
It could also listen to user-initiated clicks, and make sure we don't start a spell cast in less than X ticks after that.

Reply:
We could also make (better) predictions on the spell progress when we are not only relying on the sever send packet but also the user input.

Reply:
Added the comment.